macOS: Text cursor in HTML note is black, regardless of set Accent Color

rr-codes opened the pull request:
#### ee3805278f958bea3cdf8e44233e870f569b4b44
<pre>
macOS: Text cursor in HTML note is black, regardless of set Accent Color
<a href="https://bugs.webkit.org/show_bug.cgi?id=264189">https://bugs.webkit.org/show_bug.cgi?id=264189</a>
<a href="https://rdar.apple.com/117909679">rdar://117909679</a>

Reviewed by Aditya Keerthi and Tim Horton.

266070@main changed the behavior of the caret color to more closely follow the spec, and improve
web compat. However, this behavior change also affected HTML Notes, which sets `color` on
an ancestor of the editable div. As a result, the caret color is black. However, it should match
the accent color of the app it is in, if it sets a custom accent color.

To fix, implement the same solution as 269314@main effectively, but on macOS.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::paintCaret):
(WebCore::CaretBase::computeCaretColor):
(WebCore::CaretBase::paintCaret const):
(WebCore::DragCaretController::paintDragCaret const):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setAppUsesCustomAccentColor):
(WebCore::Page::appUsesCustomAccentColor const):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::autocorrectionReplacementMarkerColor const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::cachedAppUsesCustomAccentColor):
(WebKit::PageClientImpl::appUsesCustomAccentColor):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::reinitializeWebPage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::setAppUsesCustomAccentColor):

Canonical link: <a href="https://commits.webkit.org/270325@main">https://commits.webkit.org/270325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05854714e4e8b478220607b700a1fd3fa7ee2770

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25113 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26369 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27228 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23047 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25381 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1090 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23304 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25356 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2712 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21674 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27807 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2401 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22611 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28740 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26556 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2327 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/619 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3670 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6034 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2763 "Built successfully") | | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2659 "Built successfully") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/FrameSelection.h:
```diff
@@ -59,7 +59,7 @@ class CaretBase {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    WEBCORE_EXPORT static Color computeCaretColor(const RenderStyle& elementStyle, const Node*);
+    WEBCORE_EXPORT static Color computeCaretColor(const RenderStyle& elementStyle, const Node*, bool appUsesCustomAccentColor);
```
I don't think you need new parameters for the methods in this file? Should be able to read the bit through `Node` / `Element`.

pxlcoder commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1381,7 +1381,7 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
 Color RenderThemeIOS::autocorrectionReplacementMarkerColor(const RenderText& renderer) const
 {
-    auto caretColor = CaretBase::computeCaretColor(renderer.style(), renderer.textNode());
+    auto caretColor = CaretBase::computeCaretColor(renderer.style(), renderer.textNode(), false);
```
This looks wrong. You probably want to use `renderer.page()`. Though once the new parameter goes away, that should just work.

pxlcoder replied:
Specifically, I think the consequence of the current implementation would be mismatched autocorrection underlines and caret color in Notes.

pxlcoder commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1043,6 +1043,31 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
+        if (!usesCustomAppAccentColor && [(accentColorName = info[@"NSAppAccentColorName"]) length])
+            usesCustomAppAccentColor = !!adoptNS([NSColor colorNamed:accentColorName bundle:bundleForAccentColor]).get();
+    });
+    ALWAYS_LOG_WITH_STREAM(stream << "RR_CODES cachedAppUsesCustomAccentColor = " << usesCustomAppAccentColor);
```
😅

aprotyas reviewed (commented):
Is it possible to write an API test like `TEST(WebKit, TintColorAffectsInteractionColor)`?

aprotyas commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2389,7 +2391,7 @@ void DragCaretController::paintDragCaret(LocalFrame* frame, GraphicsContext& p,
 #if ENABLE(TEXT_CARET)
     if (m_position.deepEquivalent().deprecatedNode() && m_position.deepEquivalent().deprecatedNode()->document().frame() == frame)
-        paintCaret(*m_position.deepEquivalent().deprecatedNode(), p, paintOffset, nullptr);
+        paintCaret(*m_position.deepEquivalent().deprecatedNode(), p, paintOffset, nullptr, false);
```
Optional nit: boolean trap here. What about `enum class AppUsesCustomAccentColor : bool;` to improve readability at the call sites?

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2593,6 +2595,8 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
     bool m_useSceneKitForModel { false };
 #endif
 
+    bool m_appUsesCustomAccentColor { false };
```
Shouldn't this be in a `HAVE(APP_ACCENT_COLORS)`?

pxlcoder commented on Source/WebCore/page/Page.h:
```diff
@@ -1460,6 +1463,8 @@ class Page : public Supplementable<Page>, public CanMakeWeakPtr<Page>, public Ca
     String m_sceneIdentifier;
 #endif
+
+    bool m_appUsesCustomAccentColor { false };
```
`HAVE(APP_ACCENT_COLORS)`?

hortont424 commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1043,6 +1043,30 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
     return WebCore::colorFromCocoaColor([NSApp _effectiveAccentColor]);
 }
+
+static BOOL cachedAppUsesCustomAccentColor(void)
```
why is this not just inside PageClientImpl::appUsesCustomAccentColor?

pxlcoder commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -1893,7 +1893,9 @@ Color CaretBase::computeCaretColor(const RenderStyle& elementStyle, const Node*
     return elementStyle.colorResolvingCurrentColor(elementStyle.caretColor());
 #elif HAVE(REDESIGNED_TEXT_CURSOR)
-    if (elementStyle.hasAutoCaretColor() && !elementStyle.hasExplicitlySetColor()) {
+    auto appUsesCustomAccentColor = node && node->document().page() && node->document().page()->appUsesCustomAccentColor();
```
Doesn't this need to be behind `HAVE(APP_ACCENT_COLORS)`, so the Catalyst build doesn't break?

pxlcoder commented on Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
```diff
@@ -1043,6 +1043,30 @@ static inline NSCorrectionResponse toCorrectionResponse(AutocorrectionResponse r
+    static dispatch_once_t once;
+    static BOOL usesCustomAppAccentColor = NO;
+    dispatch_once(&once, ^{
+        // Load up the custom accent color if the Info.plist key exists, and if the app bundle's Asset Catalog contains a matching color
```
The comments here describe exactly what the code does, so not they're useful.

rr-codes committed:
macOS: Text cursor in HTML note is black, regardless of set Accent Color
https://bugs.webkit.org/show_bug.cgi?id=264189
rdar://117909679

Reviewed by Aditya Keerthi and Tim Horton.

266070@main changed the behavior of the caret color to more closely follow the spec, and improve
web compat. However, this behavior change also affected HTML Notes, which sets `color` on
an ancestor of the editable div. As a result, the caret color is black. However, it should match
the accent color of the app it is in, if it sets a custom accent color.

To fix, implement the same solution as 269314@main effectively, but on macOS.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::paintCaret):
(WebCore::CaretBase::computeCaretColor):
(WebCore::CaretBase::paintCaret const):
(WebCore::DragCaretController::paintDragCaret const):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setAppUsesCustomAccentColor):
(WebCore::Page::appUsesCustomAccentColor const):
* Source/WebCore/page/Page.h:
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::autocorrectionReplacementMarkerColor const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::cachedAppUsesCustomAccentColor):
(WebKit::PageClientImpl::appUsesCustomAccentColor):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::reinitializeWebPage):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getPlatformEditorState const):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::setAppUsesCustomAccentColor):

Canonical link: https://commits.webkit.org/270325@main
